### PR TITLE
Fix for `string.PadLeft()` and `string.Length` translation.

### PR DESCRIPTION
**Pull request by GulpakMarina**
Fix for #4799 `string.PadLeft()` translation.
Fix for #4873 `string.Length` translation.

### CODE REVIEW

File: Source/LinqToDB/DataProvider/SqlCe/Translation/SqlCeMemberTranslator.cs [hunk 1]
```diff
@@ -210,25 +210,34 @@ public class SqlCeMathMemberTranslator : MathMemberTranslatorBase
 
 		public class StringMemberTranslator : StringMemberTranslatorBase
 		{
-			public override ISqlExpression TranslateLength(ITranslationContext translationContext, TranslationFlags translationFlags, ISqlExpression value)
+			public override ISqlExpression? TranslateLength(ITranslationContext translationContext, TranslationFlags translationFlags, ISqlExpression value)
 			{
 				var factory = translationContext.ExpressionFactory;
-				return factory.Function(factory.GetDbDataType(value), "LEN", value);
+				var valueTypeString = factory.GetDbDataType(typeof(string));
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/SqlCe/Translation/SqlCeMemberTranslator.cs:216]: You should use `var valueTypeString = factory.GetDbDataType(value); `. Because string mapping get from translated vaue.

File: Source/LinqToDB/DataProvider/SqlServer/Translation/SqlServerMemberTranslator.cs [hunk 2]
```diff
@@ -172,25 +172,38 @@ public class SqlServerMathMemberTranslator : MathMemberTranslatorBase
 
 		public class StringMemberTranslator : StringMemberTranslatorBase
 		{
-			public override ISqlExpression TranslateLength(ITranslationContext translationContext, TranslationFlags translationFlags, ISqlExpression value)
+			public override ISqlExpression? TranslateLength(ITranslationContext translationContext, TranslationFlags translationFlags, ISqlExpression value)
 			{
 				var factory = translationContext.ExpressionFactory;
-				return factory.Function(factory.GetDbDataType(value), "LEN", value);
+				var valueTypeString = factory.GetDbDataType(typeof(string));
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/SqlServer/Translation/SqlServerMemberTranslator.cs:178]: Use `var valueTypeString = factory.GetDbDataType(value);`

File: Source/LinqToDB/DataProvider/SqlServer/Translation/SqlServerMemberTranslator.cs [hunk 3]
```diff
@@ -172,25 +172,38 @@ public class SqlServerMathMemberTranslator : MathMemberTranslatorBase
...
+			public override ISqlExpression? TranslateLPad(ITranslationContext translationContext, MethodCallExpression methodCall, TranslationFlags translationFlags, ISqlExpression value, ISqlExpression padding, ISqlExpression paddingChar)
 			{
 				/*
-				 * SELECT REPLICATE(paddingSymbol, padding - LEN(value)) + value
+				 * REPLICATE(paddingSymbol, padding - LEN(value)) + value
 				 */
 				var factory = translationContext.ExpressionFactory;
-				var valueType = factory.GetDbDataType(value);
-				var valueTypeInt = new DbDataType(typeof(int), DataType.Int32);
+				var valueTypeString = factory.GetDbDataType(typeof(string));
```
Comment (sdanyliv) [inline Source/LinqToDB/DataProvider/SqlServer/Translation/SqlServerMemberTranslator.cs:196]: Use `var valueTypeString = factory.GetDbDataType(value);`

File: Source/LinqToDB/Linq/Translation/StringMemberTranslatorBase.cs [hunk 4]
```diff
@@ -168,21 +168,21 @@ public StringMemberTranslatorBase()
...
-			return factory.Function(factory.GetDbDataType(value), "LENGTH", value);
+			return factory.Function(factory.GetDbDataType(typeof(int)), "LENGTH", value);
 		}
 
 		public virtual ISqlExpression? TranslateLPad(ITranslationContext translationContext, MethodCallExpression methodCall, TranslationFlags translationFlags, ISqlExpression value, ISqlExpression padding, ISqlExpression paddingChar)
 		{
 			var factory = translationContext.ExpressionFactory;
-			var valueType = factory.GetDbDataType(value);
-			return factory.Function(valueType, "LPAD", value, padding, paddingChar);
+			var valueTypeString = factory.GetDbDataType(typeof(string)); ;
```
Comment (sdanyliv) [inline Source/LinqToDB/Linq/Translation/StringMemberTranslatorBase.cs:177]: Use `var valueTypeString = factory.GetDbDataType(value);`

File: Source/LinqToDB/Linq/Translation/StringMemberTranslatorBase.cs [hunk 5]
```diff
@@ -168,21 +168,21 @@ public StringMemberTranslatorBase()
...
+			return factory.Function(valueTypeString, "LPAD", value, padding, paddingChar);
 		}
 
 		public virtual ISqlExpression? TranslatePadLeft(ITranslationContext translationContext, MethodCallExpression methodCall, TranslationFlags translationFlags, ISqlExpression value, ISqlExpression padding, ISqlExpression? paddingChar)
 		{
 			var factory = translationContext.ExpressionFactory;
 
-			var valueType = factory.GetDbDataType(value);
+			var valueType = factory.GetDbDataType(typeof(string));
```
Comment (sdanyliv) [inline Source/LinqToDB/Linq/Translation/StringMemberTranslatorBase.cs:185]: Use `var valueTypeString = factory.GetDbDataType(value);`

File: Tests/Linq/Linq/StringFunctionTests.cs [hunk 6]
```diff
@@ -293,6 +293,26 @@ public void Length([DataSources] string context)
 			}
 		}
 
+		[Test]
+		public void LengthWhiteSpace([DataSources] string context, [Values("abc ", " ", " abc ")] string stringValue)
```
Comment (MaceWindu) [inline Tests/Linq/Linq/StringFunctionTests.cs:297]: test just embed parameter values, no server-side translation tests

Comment (MaceWindu) [reply Tests/Linq/Linq/StringFunctionTests.cs:297]: also would be nice to test at least simple unicode strings without surrogates

**Review by MaceWindu — DISMISSED** [review]

**Review by MaceWindu — APPROVED** [review]
